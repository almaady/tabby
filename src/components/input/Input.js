import React from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

/**
 * @class Input - It is an html input
 */
class Input extends React.PureComponent {
  /**
   * @property {string} name -The name of the input
   * @property {string} type - The type of the input
   * @property {string} placeholder- The placeholder of the input
   * @property {func} onBlur - The function triggered when the input is onBlur
   * @property {func} onChange - The function triggered when the input value change
   * @property {bool} showError - Determines if the input should add an error style
   * @property {string} defaultValue - The Default Value of the input
   * @property {func} innerRef - The function to create a Ref for the input
   * @property {string} autoComplete - To enable autocomplete
   * @property {bool} autoFocus - Makes the input focusable when is showed on the screen
   */
  static propTypes = {
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    onBlur: PropTypes.func.isRequired,
    onChange: PropTypes.func.isRequired,
    showError: PropTypes.bool,
    defaultValue: PropTypes.string,
    innerRef: PropTypes.func,
    autoComplete: PropTypes.string,
    autoFocus: PropTypes.bool
  };

  render() {
    const {
      innerRef,
      name,
      type,
      placeholder,
      showError,
      onChange,
      onBlur,
      defaultValue,
      autoComplete,
      autoFocus
    } = this.props;
    return (
      <input
        ref={innerRef}
        autoComplete={autoComplete}
        name={name}
        type={type}
        placeholder={placeholder}
        className={classNames({ "input__error": showError }, "input")}
        onChange={onChange}
        onBlur={onBlur}
        defaultValue={defaultValue}
        autoFocus={autoFocus}
      />
    );
  }
}
Input.defaultProps = {
  defaultValue: "",
  placeholder: "",
  autoComplete: "off",
  showError: false,
  autoFocus: false,
  innerRef: () => {}

};

export default Input;
