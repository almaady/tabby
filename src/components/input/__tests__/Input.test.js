import React from "react";
import { shallow, mount } from "enzyme";
import sinon from "sinon";

import Input from "../Input";

describe("Input", () => {
  const onBlur = sinon.stub();
  const onChange = sinon.stub();
  let props;
  beforeEach(() => {
    props = {
      name: "Input",
      type: "text",
      onBlur,
      onChange,
      showError: false
    };
  });
  it("should render correctly with default state", () => {
    const component = shallow(<Input {...props} />);
    expect(component).toMatchSnapshot();
  });
  it("should add placeholder to the input", () => {
    const component = shallow(<Input {...props} placeholder="placeholder" />);
    expect(component.props().placeholder).toEqual("placeholder");
  });
  it("should call onChange when you change the input value", () => {
    const component = mount(<Input {...props} />);
    component.find("input").simulate("change");
    expect(onChange.called).toBeTruthy();
  });
  it("should call onBlur when you focus the input", () => {
    const component = mount(<Input {...props} />);
    component.find("input").simulate("blur");
    expect(onBlur.called).toBeTruthy();
  });
  it("should has class error_div with showError prop", () => {
    const component = shallow(<Input {...props} showError={true} />);
    expect(component.hasClass("input__error")).toBeTruthy();
  });
});
