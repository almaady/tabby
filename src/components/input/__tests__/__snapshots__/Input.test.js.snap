// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render correctly with default state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Input
    autoComplete="off"
    autoFocus={false}
    defaultValue=""
    innerRef={[Function]}
    name="Input"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder=""
    showError={false}
    type="text"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "autoComplete": "off",
      "autoFocus": false,
      "className": "input",
      "defaultValue": "",
      "name": "Input",
      "onBlur": [Function],
      "onChange": [Function],
      "placeholder": "",
      "type": "text",
    },
    "ref": [Function],
    "rendered": null,
    "type": "input",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "autoComplete": "off",
        "autoFocus": false,
        "className": "input",
        "defaultValue": "",
        "name": "Input",
        "onBlur": [Function],
        "onChange": [Function],
        "placeholder": "",
        "type": "text",
      },
      "ref": [Function],
      "rendered": null,
      "type": "input",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
